compiler/vm/packed_env.nim

```diff
@@ -0,0 +1,945 @@
+
+proc loadEnv*(dst: var TCtx, src: PackedEnv) =
+  ## Loads all data from `src` int `dst` for which no further/extra processing
+  ## is required. Things that are not loaded are: constants (both simple and
+  ## complex), globals, `code`, and the callback list.
+
+  # further loading requires a filled `types` list, so `loadVmTypes` has to
+  # happen first:
+  types(dst) = loadVmTypes(src)
+
+  mapList(dst.constants, src.consts, x):
+    var co = VmConstant(kind: x[0])
+    let id = x[1]
+
+    case x[0]
+    of cnstInt:     co.intVal = src.numbers[id.LitId]
+    of cnstFloat:   co.floatVal = cast[BiggestFloat](src.numbers[id.LitId])
+    of cnstString:  co.strVal = src.strings[id.LitId]
+
+    of cnstSliceListInt:   co.intSlices = loadSliceList[BiggestInt](src, id)
+    of cnstSliceListFloat: co.floatSlices = loadSliceList[BiggestFloat](src, id)
+    of cnstSliceListStr:   co.strSlices = loadSliceList[ConstantId](src, id)
+    of cnstNode: unreachable()
+
+    co
+
+  mapList(dst.functions, src.functions, x):
+    let
+      sym = src.dbgSyms[x.sym]
+      nimSym = PSym(name: PIdent(s: src.strings[sym.name]),
+                    info: src.infos[sym.info])
+    # In standalone mode, `sym` is just used to provide line information and
+    # stack trace entries
+    var f = FuncTableEntry(sym: nimSym, sig: x.sig,
+                           retValDesc: dst.types[x.t1],
+                           envParamType: dst.types[x.t2],
+                           kind: x.kind)
+    case x.kind
+    of ckDefault:
+      f.start = x.a.int
+      f.regCount = x.b.uint16
+    of ckCallback:
+      f.cbOffset = x.a.int
+
+    f
+
+  mapList(dst.debug, src.debug, x): src.infos[x.LitId]
+
+  dst.rtti = loadTypeInfos(src, dst.types)
+
+  dst.config.m.fileInfos.newSeq(src.files.len)
+  for i, f in src.files.pairs:
+    dst.config.m.fileInfos[i].fullPath = f.AbsoluteFile
+
+func init*(enc: var PackedEncoder, types: seq[PVmType]) =
+  ## Initializes the encoder. The sequence pass to `types` should not change
+  ## past this point. Else, encoding error may happen
+
+  # set up the `PVmType` -> `VmTypeId` mappings
+  # XXX: this will become unnecessary once `PVmType` is replaced by `VmTypeId`
+  for i in 0..<types.len:
+    let t = types[i]
+    enc.typeMap[t] = i.uint32
+
+  assert enc.typeMap.len == types.len
+
+func storeEnv*(enc: var PackedEncoder, dst: var PackedEnv, c: TCtx) =
+  ## Stores all relevant data provided by `c` into `dst`. Previously stored
+  ## data (except `nodes`, `numbers`, and `strings`) is thrown away. The only
+  ## parts of `dst` not touched by `storeEnv` are: `cconsts`, `globals`,
+  ## `code`, and `entryPoint`. These have to be filled in separately.
+  # store all types:
+  dst.types.newSeq(c.types.len - 1) # -1 since the invalid type is not included
+  for i in 1..<c.types.len:
+    let t = c.types[i]
+    dst.types[i - 1] = storeVmType(enc, dst, t)
+
+  # TODO: maybe make `dst.consts` a seq of indices into `dst.nodes` instead?
```

Comment:
Maybe, I think the real savings might come in when array data can be more densely stored in nodes and then const can leverage that as well. Otherwise the separation might have some benefits for now.